File: Java/src/CombinationSum.java
import java.util.ArrayList;
import java.util.List;

public class CombinationSum {
	List<List<Integer>> result;
	int[] candidates;
	int target;
	
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<List<Integer>>();
        this.candidates = candidates;
        this.target = target;
        
        List<Integer> list = new ArrayList<Integer>();
        
        dfs(list, 0, 0);
        
        return result;
    }
    
    private void dfs(List<Integer> list, int sum, int startIndex) {
    	for(int i = startIndex; i < candidates.length; i++) {
    		if(sum + candidates[i] == target) {
    			List<Integer> newList = new ArrayList<Integer>(list);
    			newList.add(candidates[i]);
    			result.add(newList);
    		}else if(sum + candidates[i] < target) {
    			list.add(candidates[i]);
    			dfs(list, sum+candidates[i], i);
    			list.remove(list.size()-1);
    		}
    	}
    }

}

File: Python/CombinationSum.py
from typing import List

class Solution:

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.res = []
        self.dfs(0, candidates, target, [], 0)
        return self.res



    def dfs(self, i, cand, target, cur, sum):
        if i >= len(cand) or sum > target:
            return
        if sum == target:
            self.res.append(cur.copy())
            return
        
        cur.append(cand[i])
        self.dfs(i, cand, target, cur, sum+cand[i])
        cur.pop()
        self.dfs(i+1, cand, target, cur, sum)

        

test = Solution()
res = test.combinationSum([2,3,5], 8)
for i in res:
    print(i)






